zebra-chain/src/orchard/shielded_data.rs

```diff
@@ -48,6 +49,12 @@ impl ShieldedData {
     pub fn nullifiers(&self) -> impl Iterator<Item = &Nullifier> {
         self.actions().map(|action| &action.nullifier)
     }
+
+    /// Collect the cm_x's for this transaction, if it contains [`Action`]s with
+    /// outputs.
+    pub fn note_commitments(&self) -> impl Iterator<Item = &pallas::Base> {
```

Comment:
One thing that's important is that the note commitments are added to the trees in the order they appear in the block, I wonder if we can/should make a note that that is guaranteed here.

Reply:
These are good. Do we have _any_ reason to think these will be returned out of order from how they appeared in the transaction? I don't _think_ so, but I'm not _certain_. Maybe @jvff @teor2345 can assuaged my concerns. Tests about taking a block (Sapling can do since the logic is ~the same) and calculating the updated tree from the commitments and checking that the computed root of the updated tree is matches what we expect may be sufficient.

Reply:
In all transaction versions:
* Zebra parses and stores note commitments in data order
* Zebra returns note commitments in data order 
* changing the order of the note commitments changes the transaction ID or authorizing data commitment, which changes the block header, which changes the block hash

So malicious remote nodes can't change the note commitment order and pass validation. And Zebra can't change the order during parsing or serialization, because existing validation or round-trip tests would fail.

So all we need to test is the order returned by each note commitment method.

Reply:
Added a test in fc229097 (expanded in 5904ee84). Not sure why I didn't think of that before, since we already have the roots in the test vectors...